contracts/upkeeps/CronUpkeep.sol

```diff
@@ -0,0 +1,311 @@
+// SPDX-License-Identifier: MIT
+
+/**
+  The Cron contract is a chainlink keepers-powered cron job runner for smart contracts.
+  The contract enables developers to trigger actions on various targets using cron
+  strings to specify the cadence. For example, a user may have 3 tasks that require
+  regular service in their dapp ecosystem:
+    1) 0xAB..CD, update(1), "0 0 * * *"     --> runs update(1) on 0xAB..CD daily at midnight
+    2) 0xAB..CD, update(2), "30 12 * * 0-4" --> runs update(2) on 0xAB..CD weekdays at 12:30
+    3) 0x12..34, trigger(), "0 * * * *"     --> runs trigger() on 0x12..34 hourly
+
+  To use this contract, a user first deploys this contract and registers it on the chainlink
+  keeper registry. Then the user adds cron jobs by following these steps:
+    1) Convert a cron string to an encoded cron spec by calling encodeCronString()
+    2) Take the encoding, target, and handler, and create a job by sending a tx to createCronJob()
+    3) Cron job is running :)
+*/
+
+pragma solidity 0.8.6;
+
+import "@chainlink/contracts/src/v0.8/dev/ConfirmedOwner.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "./KeeperBase.sol";
+import "../interfaces/KeeperCompatibleInterface.sol";
+import {CronUtility_Internal, Spec} from "../libraries/CronUtility_Internal.sol";
+import {CronUtility_External, Spec as Spec_External} from "../libraries/CronUtility_External.sol";
+import {getRevertMsg} from "../utils/utils.sol";
+
+/**
+ * @title The CronUpkeep contract
+ * @notice A keeper-compatible contract that runs various tasks on cron schedules.
+ * Users must use the encodeCronString() function to encode their cron jobs before
+ * setting them. This keeps all the string manipulation off chain and reduces gas costs.
+ */
+contract CronUpkeep is
+  KeeperCompatibleInterface,
+  KeeperBase,
+  ConfirmedOwner,
+  Pausable
+{
+  using CronUtility_Internal for Spec;
+  using CronUtility_External for Spec_External;
+  using CronUtility_External for string;
```

Comment:
Also do we need Spec_Internal and Spec_External? If CronUtility_External just uses CrontUtility_Internal.Spec, can't we have one type? Whenever you want to call CronUtility_External, you could call `CronUtility_External(s_delegate).fooBar(spec)`. I think it would be ideal to avoid ambiguity of when external calls are being used or not. Same for `string` external calls.

Reply:
yeah good call. Made that change.